### PR TITLE
[Travel] [Refactor] Create a new shared component for AddressPage

### PR DESCRIPTION
**Pull request by smelaa**
### Fixed Issues

$ https://github.com/Expensify/App/issues/41965

PROPOSAL:

### Tests

To be able to perform steps 6-7 you need to be spotnana travel beta or modify `canUseSpotnanaTravel` function from `Permissions.ts` to return always `true`.

1. Go to Account Settings > Profile and open Address Form by clicking on Address.
2. Change the address and close the modal
3. Check whether the changes were saved by reopening the modal.
4. Now change address by selecting different country and close the modal. 
5. Once again check whether the changes were saved correctly.
6. Go to Account Settings > Workspaces, choose one of the workspaces you are the admin of and open Address FOrm by clicking on Address.
7. Repeat steps 2-5.

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

Same as tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[android.webm](https://github.com/Expensify/App/assets/115273836/d8ba072a-b314-457e-829e-37c6311838cf)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[android_web.webm](https://github.com/Expensify/App/assets/115273836/98a6ed91-ed5b-43d7-bcab-ed0c8166309e)

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/115273836/09572e42-b0b2-4e6c-b7e9-7f7cc9cc4920

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/115273836/272d5a46-2abd-43b8-943e-e3c049fd5f98

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/115273836/f5a664ae-7bad-4cbb-b8c3-f7df46571303

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/115273836/273ae222-e2a9-4f91-a653-43567eab0cac

</details>

### CODE REVIEW

File: src/components/CountrySelector.tsx [hunk 1]
```diff
@@ -32,19 +33,41 @@ type CountrySelectorProps = {
 function CountrySelector({errorText = '', value: countryCode, onInputChange = () => {}, onBlur}: CountrySelectorProps, ref: ForwardedRef<View>) {
     const styles = useThemeStyles();
     const {translate} = useLocalize();
+    const route = useRoute();
+    const countryFromUrlTemp = (route?.params as Record<string, string>)?.country;
```
Comment (jnowakow) [inline src/components/CountrySelector.tsx:37]: `route.params` are typed?

File: src/components/CountrySelector.tsx [hunk 2]
```diff
@@ -32,19 +33,41 @@ type CountrySelectorProps = {
...
+    // Check if country is valid
+    const countryFromUrl = CONST.ALL_COUNTRIES[countryFromUrlTemp as keyof typeof CONST.ALL_COUNTRIES] ? countryFromUrlTemp : '';
```
Comment (jnowakow) [inline src/components/CountrySelector.tsx:39]: I think something like:
`const countryFromUrl  = CONST.ALL_COUNTRIES.find(country => country === countryFromUrlTemp)` 
looks a bit cleaner

File: src/components/CountrySelector.tsx (hunk 1 above)
Comment (blazejkustra) [reply src/components/CountrySelector.tsx:37]: How about: 
```ts
    import {SettingsNavigatorParamList} from '@libs/Navigation/types';
    import SCREENS from '@src/SCREENS';
    const route = useRoute<RouteProp<SettingsNavigatorParamList, typeof SCREENS.SETTINGS.PROFILE.ADDRESS>>();
    const countryFromUrlTemp = route?.params?.country;
    const countryFromUrl = countryFromUrlTemp && CONST.ALL_COUNTRIES[countryFromUrlTemp] ? countryFromUrlTemp : '';
```

Comment (blazejkustra) [reply src/components/CountrySelector.tsx:37]: Or better, how about moving this logic to a hook?

Comment (smelaa) [reply src/components/CountrySelector.tsx]: Done, thanks!

File: src/components/StateSelector.tsx [hunk 3]
```diff
@@ -44,7 +44,7 @@ function StateSelector(
 ) {
     const styles = useThemeStyles();
     const {translate} = useLocalize();
-    const stateFromUrl = useGeographicalStateFromRoute();
+    const stateFromUrl = useGeographicalStateAndCountryFromRoute()?.state;
```
Comment (blazejkustra) [inline src/components/StateSelector.tsx:47]: I think this looks better:
```ts
    const {state:stateFromUrl} = useGeographicalStateAndCountryFromRoute();
```

File: src/hooks/useGeographicalStateAndCountryFromRoute.ts [hunk 4]
```diff
@@ -12,12 +15,13 @@ type State = keyof typeof COMMON_CONST.STATES;
  * Example 3: Url: https://new.expensify.com/settings/profile/address Returns: undefined
  * Example 4: Url: https://new.expensify.com/settings/profile/address?state=MO-hash-a12341 Returns: MO
  */
-export default function useGeographicalStateFromRoute(stateParamName = 'state'): State | undefined {
+export default function useGeographicalStateAndCountryFromRoute(stateParamName = 'state', countryParamName = 'country'): StateAndCountry | undefined {
```
Comment (blazejkustra) [inline src/hooks/useGeographicalStateAndCountryFromRoute.ts:18]:
```suggestion
export default function useGeographicalStateAndCountryFromRoute(stateParamName = 'state', countryParamName = 'country'): StateAndCountry {
```

File: src/hooks/useGeographicalStateAndCountryFromRoute.ts [hunk 5]
```diff
@@ -1,9 +1,12 @@
 import {useRoute} from '@react-navigation/native';
 import type {ParamListBase, RouteProp} from '@react-navigation/native';
 import {CONST as COMMON_CONST} from 'expensify-common/lib/CONST';
+import CONST from '@src/CONST';
 
 type CustomParamList = ParamListBase & Record<string, Record<string, string>>;
```
Comment (blazejkustra) [inline src/hooks/useGeographicalStateAndCountryFromRoute.ts:6]: What do you think to remove CustomParamList and instead use the type from [here](https://github.com/Expensify/App/pull/41970#discussion_r1598015819) (or similar)?

File: src/components/CountrySelector.tsx [hunk 6]
```diff
@@ -32,19 +33,38 @@ type CountrySelectorProps = {
 function CountrySelector({errorText = '', value: countryCode, onInputChange = () => {}, onBlur}: CountrySelectorProps, ref: ForwardedRef<View>) {
     const styles = useThemeStyles();
     const {translate} = useLocalize();
+    const countryFromUrl = useGeographicalStateAndCountryFromRoute()?.country;
```
Comment (blazejkustra) [inline src/components/CountrySelector.tsx:36]: Same here

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/hooks/useGeographicalStateAndCountryFromRoute.ts (hunk 5 above)
Comment (smelaa) [reply src/hooks/useGeographicalStateAndCountryFromRoute.ts:6]: I have changed it a bit, what about how it is done now?

**Review by jnowakow — APPROVED** [review]
👌

File: src/hooks/useGeographicalStateAndCountryFromRoute.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import {useRoute} from '@react-navigation/native';
+import {CONST as COMMON_CONST} from 'expensify-common/lib/CONST';
+import CONST from '@src/CONST';
+
+type State = keyof typeof COMMON_CONST.STATES;
+type Country = keyof typeof CONST.ALL_COUNTRIES;
+type StateAndCountry = {state?: State; country?: Country};
+
+/**
+ * Extracts the 'state' (default) query parameter from the route/ url and validates it against COMMON_CONST.STATES, returning its ISO code or `undefined`.
```
Comment (blazejkustra) [inline src/hooks/useGeographicalStateAndCountryFromRoute.ts:10]: Let's also adjust this JSDoc 😄

**Review by blazejkustra — APPROVED** [review]

**Comment by getusha** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by getusha** [comment]
`BUG:` Workspace company address page flickers when opening
1. Open a workspace you own 
2. Profile > Company address

https://github.com/Expensify/App/assets/75031127/7a9d707c-c4da-4ce5-8bf9-4934a1791272

**Comment by getusha** [comment]
@smelaa and we've a merge conflict.

**Comment by smelaa** [comment]
@getusha the review is addressed and conflicts are solved but I cannot play [the video](https://github.com/Expensify/App/pull/41970#issuecomment-2119193832). Could you please upload it once again?

**Comment by getusha** [comment]
@smelaa 


https://github.com/Expensify/App/assets/75031127/7a9d707c-c4da-4ce5-8bf9-4934a1791272

**Comment by smelaa** [comment]
@getusha I am not able to reproduce this problem.
Chrome on Android:

https://github.com/Expensify/App/assets/115273836/69cafd06-fb52-4370-9096-f5ff3041c376


Safari on IOS:

https://github.com/Expensify/App/assets/115273836/17247704-a1f0-4873-b6f7-ca205682366f


Does it happen on main branch or different devices?

**Comment by getusha** [comment]
> Does it happen on main branch or different devices?

I can reproduce it on main, it's probably an issue from the emulator.

**Review by getusha — APPROVED** [review]

**Review by getusha — APPROVED** [review]

**Review by mountiny — DISMISSED** [review]
Changes look good to me, thanks @smelaa!

@getusha I want to make sure, have you tested also deeplinking to the country/state pages and navigating from them? thanks

**Review by mountiny — APPROVED** [review]
Thanks!

**Pull request merged**